official/recommendation/ncf.py

```diff
@@ -0,0 +1,235 @@
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import ast
+import heapq
+import math
+import os
+import sys
+import time
+
+import tensorflow as tf  # pylint: disable=g-bad-import-order
+
+import data_download
+import ncf_dataset
+import neumf_model
+import numpy as np
+from six.moves import xrange  # pylint: disable=redefined-builtin
+
+
+def evaluate_model(est_model, user_input, item_input, gt_items, top_k):
+  """Model evaluation with HR and NDCG metrics.
+
+  The evaluation protocol is to rank the test interacted item among the randomly
+  chosen 100 items that are not interacted by the user. The performance of the
+  ranked list is judged by Hit Ratio (HR) and Normalized Discounted Cumulative
+  Gain (NDCG).
+
+  For evaluation, the ranked list is truncated at 10 for both metrics. As such,
+  the HR intuitively measures whether the test item is present on the top-10
+  list, and the NDCG accounts for the position of the hit by assigning higher
+  scores to hits at top ranks. Both metrics are calculated for each test user,
+  and the average scores are reported.
+
+  Args:
+    est_model: The Estimator.
+    user_input: The user input for evaluation.
+    item_input: The item input for evaluation.
+    gt_items: The test item for HR and NDCG calculation.
+    top_k: The top-k list. It is 10 by default.
+
+  Returns:
+    hits: HR score.
+    ndcgs: NDCG score.
+  """
+  predict_input_fn = tf.estimator.inputs.numpy_input_fn(
+      x={'user_input': user_input, 'item_input': item_input},
+      batch_size=100,
+      num_epochs=1,
+      shuffle=False
+  )
+  # Get predictions
+  predictions = est_model.predict(input_fn=predict_input_fn)
+  all_predicted_scores = [p['prediction'] for p in predictions]
+
+  # Calculate HR score
+  def _get_hr(ranklist, gt_item):
+    for item in ranklist:
+      if item == gt_item:
+        return 1
+    return 0
+
+  # Calculate NDCG score
+  def _get_ndcg(ranklist, gt_item):
+    for i in range(len(ranklist)):
+      item = ranklist[i]
+      if item == gt_item:
+        return math.log(2) / math.log(i+2)
+    return 0
+
+  hits, ndcgs = [], []
+  num_users = len(gt_items)
+  step = len(user_input) // num_users  # Step for each user
+  # Evaluation on each user
+  for idx in range(num_users):
+    start = idx * step
+    end = (idx + 1) * step
+
+    items = item_input[start:end]
+    predicted_scores = all_predicted_scores[start:end]
+
+    map_item_score = {}
+    for i in range(len(items)):
+      item = items[i][0]
+      score = predicted_scores[i][0]
+      map_item_score[item] = score
+
+    # Evaluate top rank list with HR and NDCG
+    ranklist = heapq.nlargest(top_k, map_item_score, key=map_item_score.get)
+    gt_item = gt_items[idx]
+    hr = _get_hr(ranklist, gt_item)
+    ndcg = _get_ndcg(ranklist, gt_item)
+    hits.append(hr)
+    ndcgs.append(ndcg)
+
+  return hits, ndcgs
+
+
+def main(_):
+  tf.logging.set_verbosity(tf.logging.INFO)
```

Comment:
The main function is way too long. Please extract them into sub functions, like load_data, create_model, and run_training.